File: library/src/main/java/me/shkschneider/skeleton/helper/ShortcutHelper.java
package me.shkschneider.skeleton.helper;

import android.annotation.TargetApi;
import android.content.Intent;
import android.content.pm.ShortcutInfo;
import android.content.pm.ShortcutManager;
import android.graphics.drawable.Icon;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.text.TextUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

// <https://developer.android.com/reference/android/content/pm/ShortcutManager.html>
@TargetApi(AndroidHelper.API_25)
public class ShortcutHelper {

    private static final int MAX_RECOMMENDED_SHORTCUTS = 4;

    protected ShortcutHelper() {
        // Empty
    }

    public static boolean addDynamicShortcut(@NonNull final Shortcut shortcut) {
        final ShortcutManager shortcutManager = SystemServices.shortcutManager();
        if (shortcutManager.getDynamicShortcuts().size() + 1 >= MAX_RECOMMENDED_SHORTCUTS) {
            LogHelper.wtf("Lots of shortcuts");
        }
        return (! shortcutManager.isRateLimitingActive()
                && shortcutManager.addDynamicShortcuts(Collections.singletonList(shortcut.shortcutInfo())));

    }

    public static boolean setDynamicShortcuts(@NonNull final Shortcut ... shortcuts) {
        if (shortcuts.length >= MAX_RECOMMENDED_SHORTCUTS) {
            LogHelper.wtf("Lots of shortcuts");
        }
        final ShortcutManager shortcutManager = SystemServices.shortcutManager();
        if (shortcutManager.isRateLimitingActive()) {
            return false;
        }

        final List<ShortcutInfo> shortcutInfos = new ArrayList<>();
        for (final Shortcut shortcut : shortcuts) {
            shortcutInfos.add(shortcut.shortcutInfo());
        }
        return shortcutManager.setDynamicShortcuts(shortcutInfos);
    }

    public static class Shortcut {

        private final String id;
        private final int icon;
        private final String label;
        private final Intent intent;

        public Shortcut(@NonNull final String id,
                        @DrawableRes final int icon,
                        @NonNull final String label,
                        @NonNull final Intent intent) {
            this.id = id;
            this.icon = icon;
            this.label = label;
            if (TextUtils.isEmpty(intent.getAction())) {
                intent.setAction(ApplicationHelper.packageName() + "\\." + id);
            }
            this.intent = intent;
        }

        private ShortcutInfo shortcutInfo() {
            return new ShortcutInfo.Builder(ContextHelper.applicationContext(), this.id)
                    .setIcon(Icon.createWithResource(ContextHelper.applicationContext(), this.icon))
                    .setShortLabel(this.label)
                    .setIntent(this.intent)
                    .build();
        }

    }

}

File: library/src/main/java/me/shkschneider/skeleton/helper/ShortcutHelper.kt
package me.shkschneider.skeleton.helper

import android.content.Intent
import android.content.pm.ShortcutInfo
import android.graphics.drawable.Icon
import androidx.annotation.DrawableRes
import androidx.annotation.RequiresApi
import me.shkschneider.skeleton.helperx.Logger
import me.shkschneider.skeleton.helperx.SystemServices

// <https://developer.android.com/reference/android/content/pm/ShortcutManager.html>
@RequiresApi(AndroidHelper.API_25)
object ShortcutHelper {

    private val maxRecommendedShortcuts by lazy {
        SystemServices.shortcutManager()?.maxShortcutCountPerActivity ?: 4
    }

    open class Shortcut(
            private var id: String,
            @DrawableRes private var icon: Int,
            private var label: String,
            private var intent: Intent
    ) {

        init {
            if (intent.action.isNullOrEmpty()) {
                intent.action = ApplicationHelper.packageName() + "\\." + this.id
            }
        }

        fun shortcutInfo(): ShortcutInfo {
            return ShortcutInfo.Builder(ContextHelper.applicationContext(), id)
                    .setIcon(Icon.createWithResource(ContextHelper.applicationContext(), icon))
                    .setShortLabel(label)
                    .setIntent(intent)
                    .build()
        }

    }

    fun addDynamicShortcut(shortcut: Shortcut): Boolean {
        SystemServices.shortcutManager()?.let { shortcutManager ->
            val shortcutsInfo = shortcutManager.dynamicShortcuts
            if (shortcutsInfo.size + 1 >= maxRecommendedShortcuts) {
                Logger.warning("Lots of shortcuts")
            }
            return shortcutManager.isRateLimitingActive && shortcutManager.addDynamicShortcuts(listOf(shortcut.shortcutInfo()))
        }
        Logger.warning("ShortcutManager was NULL")
        return false
    }

    fun setDynamicShortcuts(vararg shortcuts: Shortcut): Boolean {
        if (shortcuts.size >= maxRecommendedShortcuts) {
            Logger.warning("Lots of shortcuts")
        }
        val shortcutManager = SystemServices.shortcutManager() ?: return false
        if (shortcutManager.isRateLimitingActive) {
            return false
        }
        val shortcutInfos = shortcuts.map { it.shortcutInfo() }
        return shortcutManager.setDynamicShortcuts(shortcutInfos)
    }

    fun removeDynamicShortcuts(): Boolean {
        val shortcutManager = SystemServices.shortcutManager() ?: return false
        shortcutManager.removeAllDynamicShortcuts()
        return true
    }

}
